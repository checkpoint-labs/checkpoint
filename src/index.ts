import Checkpoint from './checkpoint';
export { LogLevel } from './utils/logger';
export { AsyncMySqlPool } from './mysql';
export { createGetLoader } from './graphql';
export { Model } from './orm';
export * from './types';

export default Checkpoint;
