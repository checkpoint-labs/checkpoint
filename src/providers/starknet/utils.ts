import type { Abi } from 'starknet';

type StructAbi = {
  name: string;
  size: number;
  members: { name: string; type: string }[];
};

export const parseStruct = (
  type: string,
  data: string[],
  { current, structs }: { current: number; structs: Record<string, StructAbi> }
) => {
  const struct = structs[type];
  let structCurrent = current;

  return struct.members.reduce((output, field) => {
    if (structs[field.type]) {
      output[field.name] = parseStruct(field.type, data, { current: structCurrent, structs });
      structCurrent += structs[field.type].size;

      return output;
    }

    output[field.name] = data[structCurrent];
    structCurrent++;

    return output;
  }, {});
};

export const parseEvent = (abi: Abi, name: string, data: string[]): Record<string, any> => {
  const event = abi.find(el => el.name === name && el.type === 'event');
  if (!event) throw new Error('Unsupported event');

  const structs = abi
    .filter(el => el.type === 'struct')
    .reduce((acc, structAbi) => {
      acc[structAbi.name] = structAbi;
      return acc;
    }, {});

  let length = 0;
  let current = 0;

  return event.data.reduce((output, field) => {
    if (length > 0) {
      output[field.name] = data.slice(current, current + length);
      current += length;
      length = 0;

      return output;
    }

    if (structs[field.type]) {
      output[field.name] = parseStruct(field.type, data, { current, structs });
      current += structs[field.type].size;

      return output;
    }

    output[field.name] = data[current];
    if (field.name.endsWith('_len')) {
      length = parseInt(BigInt(data[current]).toString());
    }

    current++;

    return output;
  }, {});
};
