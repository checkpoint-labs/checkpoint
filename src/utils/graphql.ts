import {
  GraphQLObjectType,
  GraphQLOutputType,
  GraphQLNonNull,
  isLeafType,
  isListType,
  GraphQLScalarType
} from 'graphql';
import { jsonToGraphQLQuery } from 'json-to-graphql-query';
import pluralize from 'pluralize';

export const extendSchema = (schema: string): string => {
  return `directive @derivedFrom(field: String!) on FIELD_DEFINITION
${schema}`;
};

/**
 * Returns name of query for fetching single entity record
 *
 */
export const singleEntityQueryName = (entity: GraphQLObjectType) => entity.name.toLowerCase();

/**
 * Returns name of query for fetching multiple entity records
 *
 */
export const multiEntityQueryName = (entity: GraphQLObjectType) => {
  if (entity.name === '_Metadata') return '_metadatas';

  return pluralize(entity.name.toLowerCase());
};

/**
 * Generate sample query string based on entity object fields.
 *
 */
export const generateQueryForEntity = (entity: GraphQLObjectType): string => {
  // function to recursively build fields map
  const getObjectFields = (object: GraphQLObjectType, queryFields = {}): Record<string, any> => {
    const objectFields = object.getFields();

    Object.keys(objectFields).forEach(fieldName => {
      const rawFieldType = objectFields[fieldName].type;
      const fieldType = rawFieldType instanceof GraphQLNonNull ? rawFieldType.ofType : rawFieldType;

      if (isLeafType(fieldType)) {
        queryFields[fieldName] = true;
      } else if (isListType(fieldType)) {
        if (fieldType.ofType instanceof GraphQLScalarType) {
          queryFields[fieldName] = true;
        }
      } else {
        const childObjectFields = {};
        getObjectFields(fieldType as GraphQLObjectType, childObjectFields);
        queryFields[fieldName] = childObjectFields;
      }
    });

    return queryFields;
  };

  return jsonToGraphQLQuery(
    {
      query: {
        [multiEntityQueryName(entity)]: {
          __args: { first: 10 },
          ...getObjectFields(entity)
        }
      }
    },
    { pretty: true }
  );
};

export const getNonNullType = (type: GraphQLOutputType): GraphQLOutputType => {
  if (type instanceof GraphQLNonNull) {
    return type.ofType;
  }

  return type;
};
