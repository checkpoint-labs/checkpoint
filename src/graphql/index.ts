import { graphqlHTTP } from 'express-graphql';
import {
  GraphQLID,
  GraphQLInt,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLString
} from 'graphql';
import DataLoader from 'dataloader';
import { ResolverContextInput } from './resolvers';
import { getTableName } from '../utils/database';

/**
 * Creates getLoader function that will return existing, or create a new dataloader
 * for specific entity.
 * createGetLoader should be called per-request so each request has its own caching
 * and batching.
 */
export const createGetLoader = (context: ResolverContextInput) => {
  const loaders = {};

  return (name: string, field = 'id') => {
    const key = `${name}-${field}`;

    if (!loaders[key]) {
      loaders[key] = new DataLoader(async ids => {
        const query = context.knex
          .select('*')
          .from(getTableName(name))
          .whereIn(field, ids as string[]);

        context.log.debug({ sql: query.toQuery(), ids }, 'executing batched query');

        const results = await query;
        const resultsMap = results.reduce((acc, result) => {
          if (!acc[result[field]]) acc[result[field]] = [];

          acc[result[field]].push(result);

          return acc;
        }, {});

        return ids.map((id: any) => resultsMap[id] || []);
      });
    }

    return loaders[key];
  };
};

/**
 * Creates an graphql http handler for the query passed a parameters.
 * Returned middleware can be used with express.
 */
export default function get(
  schema: GraphQLSchema,
  context: ResolverContextInput,
  sampleQuery?: string
) {
  return graphqlHTTP(() => ({
    schema,
    context: {
      ...context,
      getLoader: createGetLoader(context)
    },
    graphiql: {
      defaultQuery: sampleQuery
    }
  }));
}

/**
 * This objects name and field maps to the values of the _metadata
 * database store
 *
 */
export const MetadataGraphQLObject = new GraphQLObjectType({
  name: '_Metadata',
  description: 'Core metadata values used internally by Checkpoint',
  fields: {
    id: { type: new GraphQLNonNull(GraphQLID), description: 'example: last_indexed_block' },
    value: { type: GraphQLString }
  }
});

/**
 * This objects name and field maps to the values of the _checkpoints
 * database store. And is used to generate entity queries for graphql
 *
 */
export const CheckpointsGraphQLObject = new GraphQLObjectType({
  name: '_Checkpoint',
  description: 'Contract and Block where its event is found.',
  fields: {
    id: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'id computed as last 5 bytes of sha256(contract+block)'
    },
    block_number: {
      type: new GraphQLNonNull(GraphQLInt)
    },
    contract_address: {
      type: new GraphQLNonNull(GraphQLString)
    }
  }
});
