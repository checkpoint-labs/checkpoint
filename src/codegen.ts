import {
  GraphQLField,
  GraphQLFloat,
  GraphQLID,
  GraphQLInt,
  GraphQLList,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLScalarType,
  GraphQLString,
  GraphQLType,
  isListType
} from 'graphql';
import pluralize from 'pluralize';
import { GqlEntityController } from './graphql/controller';
import { extendSchema } from './utils/graphql';
import { CheckpointConfig } from './types';

type TypeInfo = {
  type: string;
  initialValue: any;
};

type DecimalTypes = NonNullable<CheckpointConfig['decimal_types']>;

const DEFAULT_DECIMAL_TYPES = {
  Decimal: {
    p: 10,
    d: 2
  },
  BigDecimal: {
    p: 20,
    d: 8
  }
};

export const getTypeInfo = (
  type: GraphQLType,
  decimalTypes: DecimalTypes = DEFAULT_DECIMAL_TYPES
): TypeInfo => {
  if (type instanceof GraphQLNonNull) {
    throw new Error('Type must raw type');
  }

  switch (type) {
    case GraphQLInt:
    case GraphQLFloat:
      return { type: 'number', initialValue: 0 };
    case GraphQLString:
    case GraphQLID:
      return { type: 'string', initialValue: '' };
  }

  if (type instanceof GraphQLScalarType) {
    switch (type.name) {
      case 'BigInt':
        return { type: 'bigint', initialValue: 0 };
      case 'Boolean':
        return { type: 'boolean', initialValue: false };
      case 'Text':
        return { type: 'string', initialValue: '' };
    }

    if (type.name in decimalTypes) {
      return { type: 'string', initialValue: '0' };
    }
  }

  if (type instanceof GraphQLObjectType) {
    return { type: 'string', initialValue: '' };
  }

  if (type instanceof GraphQLList) {
    const nonNullNestedType =
      type.ofType instanceof GraphQLNonNull ? type.ofType.ofType : type.ofType;

    return { type: `${getTypeInfo(nonNullNestedType, decimalTypes).type}[]`, initialValue: [] };
  }

  throw new Error('Unknown type');
};

export const getInitialValue = (
  type: GraphQLType,
  decimalTypes: DecimalTypes = DEFAULT_DECIMAL_TYPES
) => {
  if (!(type instanceof GraphQLNonNull)) {
    return null;
  }

  return getTypeInfo(type.ofType, decimalTypes).initialValue;
};

export const getBaseType = (
  type: GraphQLType,
  decimalTypes: DecimalTypes = DEFAULT_DECIMAL_TYPES
) => {
  const nonNullType = type instanceof GraphQLNonNull ? type.ofType : type;

  return getTypeInfo(nonNullType, decimalTypes).type;
};

export const getJSType = (
  field: GraphQLField<any, any>,
  decimalTypes: DecimalTypes = DEFAULT_DECIMAL_TYPES
) => {
  const nonNullType = field.type instanceof GraphQLNonNull ? field.type.ofType : field.type;
  const isNullable = !(field.type instanceof GraphQLNonNull);
  const isList = nonNullType instanceof GraphQLList;
  const baseType = getBaseType(nonNullType, decimalTypes);

  return { isNullable, isList, baseType };
};

export const codegen = (
  schema: string,
  config: CheckpointConfig,
  format: 'typescript' | 'javascript'
) => {
  const decimalTypes = config.decimal_types || DEFAULT_DECIMAL_TYPES;
  const extendedSchema = extendSchema(schema);
  const controller = new GqlEntityController(extendedSchema, config);

  const preamble = `import { Model } from '@snapshot-labs/checkpoint';\n\n`;

  let contents = `${preamble}`;

  controller.schemaObjects.forEach((type, i, arr) => {
    const modelName = type.name;

    contents += `export class ${modelName} extends Model {\n`;
    contents += `  static tableName = '${pluralize(modelName.toLowerCase())}';\n\n`;

    contents += format === 'javascript' ? `  constructor(id) {\n` : `  constructor(id: string) {\n`;
    contents += `    super(${modelName}.tableName);\n\n`;
    controller.getTypeFields(type).forEach(field => {
      const fieldType = field.type instanceof GraphQLNonNull ? field.type.ofType : field.type;
      if (isListType(fieldType) && fieldType.ofType instanceof GraphQLObjectType) {
        return;
      }

      const rawInitialValue = getInitialValue(field.type, decimalTypes);
      const initialValue = field.name === 'id' ? 'id' : JSON.stringify(rawInitialValue);
      contents += `    this.initialSet('${field.name}', ${initialValue});\n`;
    });
    contents += `  }\n\n`;

    contents +=
      format === 'javascript'
        ? `  static async loadEntity(id) {\n`
        : `  static async loadEntity(id: string): Promise<${modelName} | null> {\n`;
    contents += `    const entity = await super.loadEntity(${modelName}.tableName, id);\n`;
    contents += `    if (!entity) return null;\n\n`;
    contents += `    const model = new ${modelName}(id);\n`;
    contents += `    model.setExists();\n\n`;
    contents += `    for (const key in entity) {\n`;
    contents += `      model.set(key, entity[key]);\n`;
    contents += `    }\n\n`;
    contents += `    return model;\n`;
    contents += `  }\n\n`;

    controller.getTypeFields(type).forEach(field => {
      const fieldType = field.type instanceof GraphQLNonNull ? field.type.ofType : field.type;
      if (isListType(fieldType) && fieldType.ofType instanceof GraphQLObjectType) {
        return;
      }

      const { isNullable, isList, baseType } = getJSType(field, decimalTypes);
      const typeAnnotation = isNullable ? `${baseType} | null` : baseType;

      contents +=
        format === 'javascript'
          ? `  get ${field.name}() {\n`
          : `  get ${field.name}(): ${typeAnnotation} {\n`;
      contents += `    return ${
        isList ? `JSON.parse(this.get('${field.name}'))` : `this.get('${field.name}')`
      };\n`;
      contents += `  }\n\n`;

      contents +=
        format === 'javascript'
          ? `  set ${field.name}(value) {\n`
          : `  set ${field.name}(value: ${typeAnnotation}) {\n`;
      contents += `    this.set('${field.name}', ${isList ? `JSON.stringify(value)` : 'value'});\n`;
      contents += `  }\n\n`;
    });

    contents = contents.slice(0, -1);
    contents += i === arr.length - 1 ? '}\n' : '}\n\n';
  });

  return contents;
};
