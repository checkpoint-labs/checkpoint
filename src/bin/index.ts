import path from 'path';
import fs from 'fs/promises';
import process from 'process';
import yargs from 'yargs/yargs';
import { hideBin } from 'yargs/helpers';
import { codegen } from '../codegen';
import { CheckpointConfig } from '../types';

const DEFAULT_CONFIG_PATH = 'src/config.json';
const DEFAULT_SCHEMA_PATH = 'src/schema.gql';
const OUTPUT_DIRECTORY = '.checkpoint';

async function generate(schemaFile: string, configFile: string, format: string) {
  if (format !== 'typescript' && format !== 'javascript') {
    throw new Error('Invalid output format');
  }

  console.log('Generating models from schema:', schemaFile);

  const cwd = process.cwd();
  const schemaFilePath = path.join(cwd, schemaFile);
  const configFilePath = path.join(cwd, configFile);

  const schema = await fs.readFile(schemaFilePath, 'utf8');
  const config: CheckpointConfig = await import(configFilePath);

  const generatedModels = codegen(schema, config, format);

  const outputFile = format === 'typescript' ? 'models.ts' : 'models.js';
  const outputPath = path.join(OUTPUT_DIRECTORY, outputFile);

  await fs.mkdir(path.join(cwd, OUTPUT_DIRECTORY), { recursive: true });
  await fs.writeFile(path.join(cwd, outputPath), generatedModels);

  console.log('Models generated to', outputPath);
}

yargs(hideBin(process.argv))
  .command(
    'generate',
    'generate models from schema',
    yargs => {
      return yargs
        .option('schema-file', {
          alias: 's',
          type: 'string',
          default: DEFAULT_SCHEMA_PATH,
          description: 'Schema file path'
        })
        .option('config-file', {
          alias: 'c',
          type: 'string',
          default: DEFAULT_CONFIG_PATH,
          description: 'Config file path'
        })
        .option('output-format', {
          alias: 'f',
          type: 'string',
          default: 'typescript',
          description: 'Output format (typescript or javascript)'
        });
    },
    async argv => {
      try {
        await generate(argv['schema-file'], argv['config-file'], argv['output-format']);
      } catch (err) {
        console.error('Error generating models:', err);
        process.exit(1);
      }
    }
  )
  .demandCommand(1, 'You need to specify a command')
  .parse();
