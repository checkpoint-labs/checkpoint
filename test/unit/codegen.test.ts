import { GraphQLObjectType, buildSchema } from 'graphql';
import { getInitialValue, getBaseType, getJSType, codegen, getTypeInfo } from '../../src/codegen';
import { extendSchema } from '../../src/utils/graphql';

const SCHEMA_SOURCE = `
scalar Id
scalar Text
scalar BigInt
scalar BigDecimal
scalar Unknown

type Space {
  id: String!
  name: String
  about: String
  controller: String!
  voting_delay: Int!
  proposal_threshold: BigInt!
  quorum: Float!
  strategies: [String]!
  strategies_nonnull: [String!]!
  proposals: [Proposal]! @derivedFrom(field: "space")
}

type Proposal {
  id: String!
  proposal_id: Int!
  space: Space!
  title: Text!
  scores_total: BigInt!
  active: Boolean!
  progress: BigDecimal!
}
`;

const schema = buildSchema(extendSchema(SCHEMA_SOURCE));
const space = schema.getType('Space') as GraphQLObjectType;
const proposal = schema.getType('Proposal') as GraphQLObjectType;
const spaceFields = space.getFields();
const proposalFields = proposal.getFields();

describe('getTypeInfo', () => {
  const simpleSchema = `scalar HugeDecimal
type Space {
  id: String!
  value: HugeDecimal
}
`;

  const customDecimalTypes = {
    HugeDecimal: {
      p: 30,
      d: 14
    }
  };

  const schema = buildSchema(extendSchema(simpleSchema));
  const space = schema.getType('Space') as GraphQLObjectType;
  const spaceFields = space.getFields();

  it('should throw when passed a wrapped type', () => {
    expect(() => getTypeInfo(spaceFields['id'].type)).toThrow();
  });

  it('should throw when passing unknown types', () => {
    expect(() => getTypeInfo(spaceFields['value'].type)).toThrow();
  });

  it('should handle non-default decimalTypes', () => {
    expect(getTypeInfo(spaceFields['value'].type, customDecimalTypes)).toEqual({
      type: 'string',
      initialValue: '0'
    });
  });
});

describe('getInitialValue', () => {
  it('should return null for nullable types', () => {
    expect(getInitialValue(spaceFields['name'].type)).toBeNull();
    expect(getInitialValue(spaceFields['about'].type)).toBeNull();
  });

  it('should return 0 for Int/Float/BigInt types', () => {
    expect(getInitialValue(spaceFields['voting_delay'].type)).toBe(0);
    expect(getInitialValue(spaceFields['proposal_threshold'].type)).toBe(0);
    expect(getInitialValue(spaceFields['quorum'].type)).toBe(0);
  });

  it('should return 0 string for BigDecimal types', () => {
    expect(getInitialValue(proposalFields['progress'].type)).toBe('0');
  });

  it('should return empty string for String/Text/Id types', () => {
    expect(getInitialValue(spaceFields['id'].type)).toBe('');
    expect(getInitialValue(spaceFields['controller'].type)).toBe('');
    expect(getInitialValue(proposalFields['title'].type)).toBe('');
  });

  it('should return false for Boolean types', () => {
    expect(getInitialValue(proposalFields['active'].type)).toBe(false);
  });

  it('should return empty array for List types', () => {
    expect(getInitialValue(spaceFields['strategies'].type)).toEqual([]);
    expect(getInitialValue(spaceFields['strategies_nonnull'].type)).toEqual([]);
  });

  it('should return empty string for object types', () => {
    expect(getInitialValue(proposalFields['space'].type)).toBe('');
  });

  it('should return "0" for BigDecimal types', () => {
    expect(getInitialValue(proposalFields['progress'].type)).toBe('0');
  });
});

describe('getBaseType', () => {
  it('should return number for Int/Float types', () => {
    expect(getBaseType(spaceFields['voting_delay'].type)).toBe('number');
    expect(getBaseType(proposalFields['proposal_id'].type)).toBe('number');
  });

  it('should return string for String/Text/Id types', () => {
    expect(getBaseType(spaceFields['id'].type)).toBe('string');
    expect(getBaseType(spaceFields['name'].type)).toBe('string');
    expect(getBaseType(proposalFields['title'].type)).toBe('string');
  });

  it('should return string for Object types', () => {
    expect(getBaseType(proposalFields['space'].type)).toBe('string');
  });

  it('should return bigint for BigInt types', () => {
    expect(getBaseType(spaceFields['proposal_threshold'].type)).toBe('bigint');
  });

  it('should return boolean for Boolean types', () => {
    expect(getBaseType(proposalFields['active'].type)).toBe('boolean');
  });

  it('should return string for BigDecimal types', () => {
    expect(getBaseType(proposalFields['progress'].type)).toBe('string');
  });

  it('should return array type for List types', () => {
    expect(getBaseType(spaceFields['strategies'].type)).toBe('string[]');
    expect(getBaseType(spaceFields['strategies_nonnull'].type)).toBe('string[]');
  });

  it('should return string for BigDecimal types', () => {
    expect(getBaseType(proposalFields['progress'].type)).toBe('string');
  });
});

describe('getJSType', () => {
  it('should detect nullable types', () => {
    expect(getJSType(spaceFields['name'])).toEqual({
      isNullable: true,
      isList: false,
      baseType: 'string'
    });
  });

  it('should detect list types', () => {
    expect(getJSType(spaceFields['strategies'])).toEqual({
      isNullable: false,
      isList: true,
      baseType: 'string[]'
    });
  });
});

describe('codegen', () => {
  const config = {
    network_node_url: '',
    sources: []
  };

  it('should generate typescript code', () => {
    expect(codegen(SCHEMA_SOURCE, config, 'typescript')).toMatchSnapshot();
  });

  it('should generate javascript code', () => {
    expect(codegen(SCHEMA_SOURCE, config, 'javascript')).toMatchSnapshot();
  });
});
