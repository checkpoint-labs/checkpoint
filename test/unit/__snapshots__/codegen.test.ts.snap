// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen should generate javascript code 1`] = `
"import { Model } from '@snapshot-labs/checkpoint';

export class Space extends Model {
  static tableName = 'spaces';

  constructor(id) {
    super(Space.tableName);

    this.initialSet('id', id);
    this.initialSet('name', null);
    this.initialSet('about', null);
    this.initialSet('controller', "");
    this.initialSet('voting_delay', 0);
    this.initialSet('proposal_threshold', 0);
    this.initialSet('quorum', 0);
    this.initialSet('strategies', []);
    this.initialSet('strategies_nonnull', []);
  }

  static async loadEntity(id) {
    const entity = await super.loadEntity(Space.tableName, id);
    if (!entity) return null;

    const model = new Space(id);
    model.setExists();

    for (const key in entity) {
      model.set(key, entity[key]);
    }

    return model;
  }

  get id() {
    return this.get('id');
  }

  set id(value) {
    this.set('id', value);
  }

  get name() {
    return this.get('name');
  }

  set name(value) {
    this.set('name', value);
  }

  get about() {
    return this.get('about');
  }

  set about(value) {
    this.set('about', value);
  }

  get controller() {
    return this.get('controller');
  }

  set controller(value) {
    this.set('controller', value);
  }

  get voting_delay() {
    return this.get('voting_delay');
  }

  set voting_delay(value) {
    this.set('voting_delay', value);
  }

  get proposal_threshold() {
    return this.get('proposal_threshold');
  }

  set proposal_threshold(value) {
    this.set('proposal_threshold', value);
  }

  get quorum() {
    return this.get('quorum');
  }

  set quorum(value) {
    this.set('quorum', value);
  }

  get strategies() {
    return JSON.parse(this.get('strategies'));
  }

  set strategies(value) {
    this.set('strategies', JSON.stringify(value));
  }

  get strategies_nonnull() {
    return JSON.parse(this.get('strategies_nonnull'));
  }

  set strategies_nonnull(value) {
    this.set('strategies_nonnull', JSON.stringify(value));
  }
}

export class Proposal extends Model {
  static tableName = 'proposals';

  constructor(id) {
    super(Proposal.tableName);

    this.initialSet('id', id);
    this.initialSet('proposal_id', 0);
    this.initialSet('space', "");
    this.initialSet('title', "");
    this.initialSet('scores_total', 0);
    this.initialSet('active', false);
    this.initialSet('progress', "0");
  }

  static async loadEntity(id) {
    const entity = await super.loadEntity(Proposal.tableName, id);
    if (!entity) return null;

    const model = new Proposal(id);
    model.setExists();

    for (const key in entity) {
      model.set(key, entity[key]);
    }

    return model;
  }

  get id() {
    return this.get('id');
  }

  set id(value) {
    this.set('id', value);
  }

  get proposal_id() {
    return this.get('proposal_id');
  }

  set proposal_id(value) {
    this.set('proposal_id', value);
  }

  get space() {
    return this.get('space');
  }

  set space(value) {
    this.set('space', value);
  }

  get title() {
    return this.get('title');
  }

  set title(value) {
    this.set('title', value);
  }

  get scores_total() {
    return this.get('scores_total');
  }

  set scores_total(value) {
    this.set('scores_total', value);
  }

  get active() {
    return this.get('active');
  }

  set active(value) {
    this.set('active', value);
  }

  get progress() {
    return this.get('progress');
  }

  set progress(value) {
    this.set('progress', value);
  }
}
"
`;

exports[`codegen should generate typescript code 1`] = `
"import { Model } from '@snapshot-labs/checkpoint';

export class Space extends Model {
  static tableName = 'spaces';

  constructor(id: string) {
    super(Space.tableName);

    this.initialSet('id', id);
    this.initialSet('name', null);
    this.initialSet('about', null);
    this.initialSet('controller', "");
    this.initialSet('voting_delay', 0);
    this.initialSet('proposal_threshold', 0);
    this.initialSet('quorum', 0);
    this.initialSet('strategies', []);
    this.initialSet('strategies_nonnull', []);
  }

  static async loadEntity(id: string): Promise<Space | null> {
    const entity = await super.loadEntity(Space.tableName, id);
    if (!entity) return null;

    const model = new Space(id);
    model.setExists();

    for (const key in entity) {
      model.set(key, entity[key]);
    }

    return model;
  }

  get id(): string {
    return this.get('id');
  }

  set id(value: string) {
    this.set('id', value);
  }

  get name(): string | null {
    return this.get('name');
  }

  set name(value: string | null) {
    this.set('name', value);
  }

  get about(): string | null {
    return this.get('about');
  }

  set about(value: string | null) {
    this.set('about', value);
  }

  get controller(): string {
    return this.get('controller');
  }

  set controller(value: string) {
    this.set('controller', value);
  }

  get voting_delay(): number {
    return this.get('voting_delay');
  }

  set voting_delay(value: number) {
    this.set('voting_delay', value);
  }

  get proposal_threshold(): bigint {
    return this.get('proposal_threshold');
  }

  set proposal_threshold(value: bigint) {
    this.set('proposal_threshold', value);
  }

  get quorum(): number {
    return this.get('quorum');
  }

  set quorum(value: number) {
    this.set('quorum', value);
  }

  get strategies(): string[] {
    return JSON.parse(this.get('strategies'));
  }

  set strategies(value: string[]) {
    this.set('strategies', JSON.stringify(value));
  }

  get strategies_nonnull(): string[] {
    return JSON.parse(this.get('strategies_nonnull'));
  }

  set strategies_nonnull(value: string[]) {
    this.set('strategies_nonnull', JSON.stringify(value));
  }
}

export class Proposal extends Model {
  static tableName = 'proposals';

  constructor(id: string) {
    super(Proposal.tableName);

    this.initialSet('id', id);
    this.initialSet('proposal_id', 0);
    this.initialSet('space', "");
    this.initialSet('title', "");
    this.initialSet('scores_total', 0);
    this.initialSet('active', false);
    this.initialSet('progress', "0");
  }

  static async loadEntity(id: string): Promise<Proposal | null> {
    const entity = await super.loadEntity(Proposal.tableName, id);
    if (!entity) return null;

    const model = new Proposal(id);
    model.setExists();

    for (const key in entity) {
      model.set(key, entity[key]);
    }

    return model;
  }

  get id(): string {
    return this.get('id');
  }

  set id(value: string) {
    this.set('id', value);
  }

  get proposal_id(): number {
    return this.get('proposal_id');
  }

  set proposal_id(value: number) {
    this.set('proposal_id', value);
  }

  get space(): string {
    return this.get('space');
  }

  set space(value: string) {
    this.set('space', value);
  }

  get title(): string {
    return this.get('title');
  }

  set title(value: string) {
    this.set('title', value);
  }

  get scores_total(): bigint {
    return this.get('scores_total');
  }

  set scores_total(value: bigint) {
    this.set('scores_total', value);
  }

  get active(): boolean {
    return this.get('active');
  }

  set active(value: boolean) {
    this.set('active', value);
  }

  get progress(): string {
    return this.get('progress');
  }

  set progress(value: string) {
    this.set('progress', value);
  }
}
"
`;
